File: src/main/java/com/readit/service/CategoryService.java
package com.readit.service;

import com.readit.entity.Category;

import java.util.List;
import java.util.Set;

public interface CategoryService {

    List<Category> getAll();

    Category getById(Long id);

    Set<Category> getBookCategories(Long bookId);

    List<Category> getRootCategories();

    List<Category> getAscendants(Long id);

    List<Category> getDescendants(Long id);
}

File: src/main/webapp/resources/app/CategoryService.js
mainModule.service('CategoryService', ['$http', function ($http) {
    return {
        getRootCategories: function () {
            return $http.get("getRootCategories")
                .then(function (response) {
                    return response.data;
                });
        },

        getBookCategories: function (bookId) {
            return $http.get("getBookCategories/" + bookId)
                .then(function (response) {
                    return response.data;
                });
        },

        getCategoryWithChildren: function (categoryId) {
            return $http.get("getCategoryWithChildren/" + categoryId)
                .then(function (response) {
                    return response.data;
                });
        },

        getCategoryWithParent: function (categoryId) {
            return $http.get("getCategoryWithParent/" + categoryId)
                .then(function (response) {
                    return response.data;
                });
        }

    }
}]);